Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/debuglog/DebugLogger.kt

```diff
@@ -0,0 +1,161 @@
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.yield
+import timber.log.Timber
+import java.io.File
+
+@SuppressLint("LogNotTimber")
+@Suppress("BlockingMethodInNonBlockingContext")
+object DebugLogger : DebugLoggerBase() {
+
+    private val scope = DebugLoggerScope
+    private lateinit var context: Context
+
+    private val debugDir by lazy {
+        File(context.cacheDir, "debuglog").also {
+            if (!it.exists()) it.mkdir()
+        }
+    }
+
+    private val triggerFile by lazy { File(debugDir, "debug.trigger") }
+
+    internal val runningLog by lazy { File(debugDir, "debug.log") }
+    val sharedDirectory by lazy { File(debugDir, "shared") }
+
+    private val mutex = Mutex()
+
+    private var logJob: Job? = null
+    private var logTree: DebugLogTree? = null
+    private var isDaggerReady = false
+
+    fun init(application: Application) {
+        context = application
+
+        if (triggerFile.exists()) {
+            Timber.tag(TAG).i("Trigger file exists, starting debug log.")
+            runBlocking { start() }
+        }
+    }
+
+    /**
+     * To censor unique data, we need to actually know what to censor.
+     * So we buffer log statements until Dagger is ready
+     */
+    fun setInjectionIsReady(component: ApplicationComponent) {
+        Timber.tag(TAG).i("setInjectionIsReady()")
+        component.inject(this)
+        isDaggerReady = true
+    }
+
+    val isLogging: Boolean
+        get() = logJob?.isActive == true
+
+    suspend fun start(): Unit = mutex.withLock {
+        Timber.tag(TAG).d("start()")
+
+        if (isLogging) {
+            Timber.tag(TAG).w("Ignoring start(), already running.")
+            return@withLock
+        }
+
+        logJob?.cancel()
+        logTree?.let { Timber.uproot(it) }
+
+        DebugLogTree().apply {
+            Timber.plant(this)
+            logTree = this
+
+            if (!runningLog.exists()) {
+                runningLog.parentFile?.mkdirs()
+                if (runningLog.createNewFile()) {
+                    Timber.tag(TAG).i("Log file didn't exist and was created.")
+                }
+            }
+
+            logJob = scope.launch {
+                try {
+                    logLines.collect { rawLine ->
+                        while (!isDaggerReady) {
+                            yield()
+                        }
+                        val censoredLine = bugCensors.get().mapNotNull { it.checkLog(rawLine) }.firstOrNull()
+                        appendLogLine(censoredLine ?: rawLine)
+                    }
+                } catch (e: CancellationException) {
+                    Timber.tag(TAG).i("Logging was canceled.")
+                } catch (e: Exception) {
+                    Log.e(TAG, "Failed to call appendLogLine(...)", e)
+                }
+            }
+        }
+
+        if (!triggerFile.exists()) {
+            Timber.tag(TAG).i("Trigger file created.")
+            triggerFile.createNewFile()
```

Comment:
Shouldn't we catch the exception?

Reply:
Good point. Especially the init on app start.
I've changed it to catch the exception on app init, and to display the exception if it happens on the debug log screen.